### ISSUE TITLE
Build JUnit 5 using JDK 14

### ISSUE THREAD

**Issue by sormuras** [junit-team/junit5#1936]
Follow-up from #1934 

> Groovy-
> ```
> Task :junit-jupiter-engine:compileTestGroovy FAILED
>
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':junit-jupiter-engine:compileTestGroovy'.
> > BUG! exception in phase 'semantic analysis' in source unit
>  '...\junit-jupiter-engine\src\test\groovy\
>    org\junit\jupiter\api\GroovyAssertEqualsTests.groovy'
> Unsupported class file major version 58
> ```
> ... and Mockito-based tests...
> ```
> org.junit.jupiter.params.provider.MethodArgumentsProviderTests
>   providesArgumentsUsingIterator() FAILED
>     org.mockito.exceptions.base.MockitoException:
>     Cannot call abstract real method on java object!
>     Calling real methods is only possible when mocking non abstract method.
> ```
> ...failed to compile/run on JDK 14. JDK 11, 12 and 13 are okay.

## Deliverables

- [x] Upgrade to Gradle/Groovy version that supports JDK 14
- [x] Upgrade to Mockito version that supports JDK 14
- [ ] Add Azure build for JDK 14

**Comment by sormuras** [comment]
@mockito 3.0.0 now supports JDK 14, correct @raphw?

**Comment by raphw** [comment]
No, not yet. It really is only possible once ASM supports it properly where ASM is still in beta but I am currently working on a release with experimental support (`-Dnet.bytebuddy.experimental=true`) which should come out today (*1.9.14*).

**Comment by raphw** [comment]
The new Byte Buddy version is released now, you can include it manually and set the experimental option and it should work.

**Comment by sormuras** [comment]
#1063 was just closed by upgrading to a recent Groovy 3 beta release.

**Comment by marcphilipp** [comment]
I've enabled byte-buddy's experimental option in 7b2ed227c276bcb2194a6224dbb604ebbbabd736 and was able to run a local build on OpenJDK 14-valhalla+2-18.

However, there's still no openjdk Docker image that we could use for Azure.

**Comment by sormuras** [comment]
Trying with @shipilev 's docker image at https://github.com/junit-team/junit5/compare/issues/1936-jdk-14

Only the correct path to `JAVA_HOME` is missing...

**Comment by shipilev** [comment]
`JAVA_HOME` should be fixed now, along with other environment variables in `shipilev:openjdk-latest-*`. Please try again.

**Comment by sormuras** [comment]
@marcphilipp Please take over. 😉

**Comment by marcphilipp** [comment]
Azure Pipelines seems to create its own users and run commands in the container using it: https://dev.azure.com/junit-team/junit5/_build/results?buildId=508&view=logs&jobId=1aad50ee-49f1-5da3-3759-715a0d68312e&taskId=51f9df4a-090a-49c5-b774-b94af3795e7b&lineStart=47&lineEnd=48&colStart=1&colEnd=1

Which doesn't work with this image because Java is installed in `/root/jdk`.

**Comment by shipilev** [comment]
Right. Then `/opt/jdk`, like in updated `shipilev/openjdk:14-fastdebug`?

**Comment by sormuras** [comment]
Still `ERROR: JAVA_HOME is set to an invalid directory: /opt/jdk`

https://dev.azure.com/junit-team/junit5/_build/results?buildId=509

**Comment by marcphilipp** [comment]
I think the file permissions are still too restrictive:

```
# ls -al /opt/jdk/
total 32
drwxr-xr-x  7 root root 4096 Jul 15 13:05 .
drwxr-xr-x  1 root root 4096 Jul 15 13:05 ..
drwx------  2 root root 4096 Jul 14 21:31 bin
drwx------  5 root root 4096 Jul 14 21:31 conf
drwx------  3 root root 4096 Jul 14 21:31 include
drwx------ 72 root root 4096 Jul 14 21:31 legal
drwx------  5 root root 4096 Jul 14 21:31 lib
-rw-------  1 root root 1184 Jul 14 21:31 release
```

**Comment by shipilev** [comment]
Grrr. Fixed to `a+rx`: 
```
$ docker run --rm -it shipilev/openjdk:14-fastdebug ls -la /opt/jdk/
total 32
drwxr-xr-x  7 root root 4096 Jul 15 15:31 .
drwxr-xr-x  3 root root 4096 Jul 15 15:31 ..
drwxr-xr-x  2 root root 4096 Jul 14 21:31 bin
drwxr-xr-x  5 root root 4096 Jul 14 21:31 conf
drwxr-xr-x  3 root root 4096 Jul 14 21:31 include
drwxr-xr-x 72 root root 4096 Jul 14 21:31 legal
drwxr-xr-x  5 root root 4096 Jul 14 21:31 lib
-rwxr-xr-x  1 root root 1184 Jul 14 21:31 release
```

**Comment by sormuras** [comment]
Restarted at https://dev.azure.com/junit-team/junit5/_build/results?buildId=510 ...

**Comment by sormuras** [comment]
... build has started! 🚀 

```
------------------------------------------------------------
Gradle 5.5.1
------------------------------------------------------------
Build time:   2019-07-10 20:38:12 UTC
Revision:     3245f748c7061472da4dc184991919810f7935a5
Kotlin:       1.3.31
Groovy:       2.5.4
Ant:          Apache Ant(TM) version 1.9.14 compiled on March 12 2019
JVM:          14-testing (Aleksey Shipilev 14-testing+0-builds.shipilev.net-openjdk-jdk-b935-20190714-jdk-145)
OS:           Linux 4.15.0-1047-azure amd64
```

**Comment by marcphilipp** [comment]
FWIW I submitted a PR to the repo that builds the official Docker images:
https://github.com/docker-library/openjdk/pull/343

**Comment by sormuras** [comment]
...the build from above "failed" due to timing issue:

```
Failures (1):
  JUnit Jupiter:KotlinAssertionsDemo:timeout not exceeded testing()
    MethodSource [className = 'example.KotlinAssertionsDemo', methodName = 'timeout not exceeded testing', methodParameterTypes = '']
    => org.opentest4j.AssertionFailedError: execution exceeded timeout of 100 ms by 23 ms
       org.junit.jupiter.api.AssertionUtils.fail(AssertionUtils.java:39)
```

TL;DR ... it builds on JDK 14-ea.

**Comment by sormuras** [comment]
> FWIW I submitted a PR to the repo that builds the official Docker images:
> [docker-library/openjdk#343](https://github.com/docker-library/openjdk/pull/343)

...  build is running on that image as well: https://dev.azure.com/junit-team/junit5/_build/results?buildId=512

**Comment by shipilev** [comment]
> ...the build from above "failed" due to timing issue:

Yes, do note you are running "fastdebug" build: it is slower than default "release", and that is the cost of running more JVM checks.
